### PR TITLE
gh-112070: make `functools.lru_cache` threadsafe in --disable-gil build

### LINKED ISSUE #112070: Make `functools.lru_cache` thread-safe in `--disable-gil` builds
# Feature or enhancement

We should make `functools.lru_cache` thread-safe in the `--disable-gil` builds.

For context, here is the commit from the `nogil-3.12` fork: https://github.com/colesbury/nogil-3.12/commit/041a08e339

**NOTES** (differences in 3.13 from `nogil-3.12`):
1) No need for an extra mutex in `lru_cache_object`; every PyObject has a mutex in the `--disable-gil` builds
2) For `_functools__lru_cache_wrapper_cache_info_impl` and `_functools__lru_cache_wrapper_cache_clear_impl` we should instead use the `@critical_section` Arugment Clinic directive. This will be simpler and require fewer changes to the code. `lru_cache_call` still needs explicit calls to the critical section API.

### Linked PRs
* gh-112111

### PR DESCRIPTION
**Pull request by wanderxjtu**
Fixes #112070 

* Issue: gh-112070

### CODE REVIEW

File: Modules/_functoolsmodule.c [hunk 1]
```diff
@@ -6,6 +6,7 @@
 #include "pycore_object.h"        // _PyObject_GC_TRACK
 #include "pycore_pystate.h"       // _PyThreadState_GET()
 #include "pycore_tuple.h"         // _PyTuple_ITEMS()
+#include "pycore_critical_section.h"  // Py_BEGIN_CRITICAL_SECTION
```
Comment (colesbury) [inline Modules/_functoolsmodule.c]: The includes are mostly sorted. Would you please put this between pycore_call.h and pycore_dict.h?

File: Modules/_functoolsmodule.c [hunk 2]
```diff
@@ -1274,7 +1275,11 @@ lru_cache_dealloc(lru_cache_object *obj)
 static PyObject *
 lru_cache_call(lru_cache_object *self, PyObject *args, PyObject *kwds)
 {
-    return self->wrapper(self, args, kwds);
+    PyObject* result;
```
Comment (colesbury) [inline Modules/_functoolsmodule.c]: Pointers are usually formatted as `PyObject *result;` in CPython

**Review by colesbury — COMMENTED** [review]
Thanks @wanderxjtu! I left a few comments inline. I don't think this change needs a NEWS entry, so I added the "skip news" label and removed the NEWS file.

**Review by colesbury — APPROVED** [review]
Thanks @wanderxjtu!

@corona10, would you please review this PR?

**Comment by corona10** [comment]
I will take a look

**Review by corona10 — APPROVED** [review]
lgtm

**Pull request merged**